### PR TITLE
Maybe improvements

### PR DESCRIPTION
**Pull request by jkozlowski**
## After this PR
==COMMIT_MSG==

==COMMIT_MSG==

### CODE REVIEW

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java [hunk 1]
```diff
@@ -159,14 +148,57 @@ private void preSubscribeLogging() {
...
+
+            REFRESHABLE_CLEANER.register(child, disposable::dispose);
+
+            return child;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    private <R> SettableRefreshable<R> createSettableChild(R initialChildValue) {
```
Comment (carterkozak) [inline src/main/java/com/palantir/refreshable/DefaultRefreshable.java:169]: This is interesting, and raises something I've been thinking about:

I don't think we necessarily care that the children are settable refreshables, in fact we should probably wrap the returned refreshable with a thin shim to remove implementation type information (we would never want someone to cast to settablerefreshable and update a mapped refreshable!). For a separate PR/feature though!
We happen to use the same codepath out of convenience, so I think the previous `DefaultRefreshable` type (for private implementation details) was reasonable.

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java [hunk 2]
```diff
@@ -159,14 +148,57 @@ private void preSubscribeLogging() {
...
+        MappingSubscriber<? super T, R> mapSubscriber = new MappingSubscriber<>(function, childRef);
+        return subscribeToSelf(mapSubscriber, _ignored -> {});
+    }
+
     @GuardedBy("readLock")
-    private Disposable subscribeToSelf(Consumer<? super T> subscriber) {
+    private <C extends NoThrowSubscriber<? super T>> Disposable subscribeToSelf(
+            C subscriber, DisposeListener<C> disposeListener) {
         preSubscribeLogging();
+        // The subscriber promises not to throw, so there isn't a way we could have a hanging
```
Comment (carterkozak) [inline src/main/java/com/palantir/refreshable/DefaultRefreshable.java:184]: A hanging what? Comment left _me_ hanging! ;-)

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java [hunk 3]
```diff
@@ -159,14 +148,57 @@ private void preSubscribeLogging() {
...
+                },
+                subscriber);
     }
 
+    interface DisposeListener<C extends Consumer<?>> {
+        void disposed(C consumer);
+    }
+
+    // Marker interface for documentation: if you implement this, you must promise not to throw!
+    interface NoThrowSubscriber<T> extends Consumer<T> {}
```
Comment (carterkozak) [inline src/main/java/com/palantir/refreshable/DefaultRefreshable.java:200]: Should we use a concrete delegating type which catches exceptions instead? That way we don't have to trust that implementations follow the contract -- they're forced to. It means we can't accidentally use method references to convert unsafe consumers.

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java [hunk 4]
```diff
@@ -178,42 +210,26 @@ private Disposable subscribeToSelf(Consumer<? super T> subscriber) {
...
-    private static final class DefaultDisposable implements Disposable {
-        private final Set<? extends Consumer<?>> subscribers;
-        private final Consumer<?> subscriber;
+    private static final class NotifyingDisposable<C extends Consumer<?>> implements Disposable {
+        private final DisposeListener<C> listener;
+        private final C subscriber;
 
-        DefaultDisposable(Set<? extends Consumer<?>> subscribers, Consumer<?> subscriber) {
-            this.subscribers = subscribers;
+        NotifyingDisposable(DisposeListener<C> listener, C subscriber) {
```
Comment (carterkozak) [inline src/main/java/com/palantir/refreshable/DefaultRefreshable.java:217]: Per comment on the other PR, I'd prefer to keep this as is and avoid additional lambda indirection. I think it makes the algorithm more mechanically obvious (harder to accidentally capture object references in future updates).

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java (hunk 2 above)
Comment (jkozlowski) [reply src/main/java/com/palantir/refreshable/DefaultRefreshable.java:184]: heh, clearly was something more important here...

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java (hunk 3 above)
Comment (jkozlowski) [reply src/main/java/com/palantir/refreshable/DefaultRefreshable.java:200]: Yes, except that existing ones already catch. But that works for me even more.

File: src/main/java/com/palantir/refreshable/DefaultRefreshable.java (hunk 4 above)
Comment (jkozlowski) [reply src/main/java/com/palantir/refreshable/DefaultRefreshable.java:217]: Cool, in that case we defo want a small class for the other lambda.

**Pull request closed without merging**